pandasai/smart_datalake/__init__.py

```diff
@@ -501,36 +355,96 @@ def chat(self, query: str, output_type: Optional[str] = None):
-
-        result = self._query_exec_tracker.execute_func(
-            self._response_parser.parse, result
-        )
+        self.update_intermediate_value_post_pipeline_execution(pipeline_context)
 
         self._query_exec_tracker.success = True
 
         self._query_exec_tracker.publish()
 
         return result
 
-    def _add_result_to_memory(self, result: dict):
+    def prepare_context_for_smart_datalake_pipeline(
+        self, query: str, output_type: Optional[str] = None
+    ) -> PipelineContext:
         """
-        Add the result to the memory.
+        Prepare Pipeline Context to intiate Smart Data Lake Pipeline.
 
         Args:
-            result (dict): The result to add to the memory
+            query (str): Query to run on the dataframe
+            output_type (Optional[str]): Add a hint for LLM which
+                type should be returned by `analyze_data()` in generated
+                code. Possible values: "number", "dataframe", "plot", "string":
+                    * number - specifies that user expects to get a number
+                        as a response object
+                    * dataframe - specifies that user expects to get
+                        pandas/polars dataframe as a response object
+                    * plot - specifies that user expects LLM to build
+                        a plot
+                    * string - specifies that user expects to get text
+                        as a response object
+                If none `output_type` is specified, the type can be any
+                of the above or "text".
+
+        Returns:
+            PipelineContext: The Pipeline Context to be used by Smart Data Lake Pipeline.
+        """
+
+        self._query_exec_tracker.start_new_track()
+
+        self.logger.log(f"Question: {query}")
+        self.logger.log(f"Running PandasAI with {self._llm.type} LLM...")
+
+        self._assign_prompt_id()
+
+        self._query_exec_tracker.add_query_info(
+            self._conversation_id, self._instance, query, output_type
+        )
+
+        self._query_exec_tracker.add_dataframes(self._dfs)
+
+        self._memory.add(query, True)
+
+        output_type_helper = output_type_factory(output_type, logger=self.logger)
+        viz_lib_helper = viz_lib_type_factory(self._viz_lib, logger=self.logger)
+
+        pipeline_context = PipelineContext(
+            dfs=self.dfs,
+            config=self.config,
+            memory=self.memory,
+            cache=self.cache,
+            query_exec_tracker=self._query_exec_tracker,
+        )
+        pipeline_context.add_intermediate_value(
+            "output_type_helper", output_type_helper
+        )
+        pipeline_context.add_intermediate_value("viz_lib_helper", viz_lib_helper)
+        pipeline_context.add_intermediate_value(
+            "last_code_generated", self._last_code_generated
+        )
+        pipeline_context.add_intermediate_value("get_prompt", self._get_prompt)
```

Comment:
@milind-sinaptik This functionality of get_prompt can be moved to logic unit that take cares of prompt generation. This way we can completely decouple it from the SmartDatalake.